Plugins/Mute/Plugin.cs

```diff
@@ -7,97 +7,203 @@ namespace Mute;
+                            .FormatExt(muteMetaJoin.Reason));
+                        break;
+                    case MuteState.Unmuted:
+                    default:
+                        break;
+                }
+
+                break;
+            case GameEvent.EventType.Say:
+                var muteMetaSay = await MuteManager.GetCurrentMuteState(gameEvent.Origin);
+                switch (muteMetaSay.MuteState)
+                {
+                    case MuteState.Muted:
+                        gameEvent.Origin.Tell(Utilities.CurrentLocalization
+                            .LocalizationIndex["PLUGINS_MUTE_REMAINING_TIME"].FormatExt(
+                                muteMetaSay.Expiration is not null
+                                    ? muteMetaSay.Expiration.Value.HumanizeForCurrentCulture()
+                                    : Utilities.CurrentLocalization.LocalizationIndex["PLUGINS_MUTE_NEVER"],
+                                muteMetaSay.Reason));
+                        break;
+                    case MuteState.Unmuting:
+                    case MuteState.Unmuted:
+                    default:
+                        break;
+                }
+
+                break;
+            case GameEvent.EventType.Update:
+                // Get correct EFClient object
+                var client = server.GetClientsAsList().FirstOrDefault(x => x.NetworkId == gameEvent.Origin.NetworkId);
+                if (client == null) break;
+
+                // Check if user has any meta set, else ignore (unmuted)
+                var muteMetaUpdate = await MuteManager.GetCurrentMuteState(client);
+                switch (muteMetaUpdate.MuteState)
+                {
+                    case MuteState.Muted:
+                        if (!muteMetaUpdate.CommandExecuted)
+                        {
+                            await server.ExecuteCommandAsync($"muteClient {client.ClientNumber}");
+                            muteMetaUpdate.CommandExecuted = true;
+                        }
+
+                        await IsExpiredMute(client, server);
                         break;
                     case MuteState.Unmuting:
-                        await server.ExecuteCommandAsync($"unmute {gameEvent.Origin.ClientNumber}");
-                        await DataManager.WritePersistentData(gameEvent.Origin, MuteState.Unmuted);
                         break;
                     case MuteState.Unmuted:
+                        if (!muteMetaUpdate.CommandExecuted)
+                        {
+                            await server.ExecuteCommandAsync($"unmute {client.ClientNumber}");
+                            muteMetaUpdate.CommandExecuted = true;
+                        }
+
                         break;
                 }
 
                 break;
         }
     }
 
-    public Task OnLoadAsync(IManager manager)
+    private async Task<bool> IsExpiredMute(EFClient client, Server server)
```

Comment:
It doesn't make sense for this method to both return the state of expiration and also unmute